lib/core/proxy.c

```diff
@@ -398,11 +442,35 @@ static inline void on_websocket_upgrade(struct rp_generator_t *self, h2o_timeout
     h2o_req_t *req = self->src_req;
     h2o_socket_t *sock = h2o_http1client_steal_socket(self->client);
     h2o_buffer_consume(&sock->input, rlen); // trash data after stealing sock.
-    struct rp_ws_upgrade_info_t *info = h2o_mem_alloc(sizeof(*info));
-    info->upstream_sock = sock;
-    info->timeout = timeout;
-    info->ctx = req->conn->ctx;
-    h2o_http1_upgrade(req, NULL, 0, on_websocket_upgrade_complete, info);
+
+    if (req->version < 0x200) {
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_UPGRADE, NULL, H2O_STRLIT("websocket"));
+        struct rp_ws_upgrade_info_t *info = h2o_mem_alloc(sizeof(*info));
+        info->upstream_sock = sock;
+        info->timeout = timeout;
+        info->ctx = req->conn->ctx;
+        h2o_http1_upgrade(req, NULL, 0, on_websocket_upgrade_complete, info);
+    } else {
+        assert(self->websocket_key != NULL);
+        ssize_t cursor = h2o_find_header_by_str(&req->res.headers, H2O_STRLIT("sec-websocket-accept"), -1);
+        if (cursor == -1)
+            goto OnInvalidResponse;
+
+        h2o_header_t accept_header = req->res.headers.entries[cursor];
+        h2o_delete_header(&req->res.headers, cursor);
+        char accept_key[29];
+        create_websocket_accept_key(accept_key, self->websocket_key);
+        if (!h2o_memis(accept_key, strlen(accept_key), accept_header.value.base, accept_header.value.len))
+            goto OnInvalidResponse;
+
+        req->res.status = 200;
+        h2o_http2_tunnel(req, &h2o_tunnel_socket_endpoint_callbacks, sock, timeout);
```

Comment:
I would suggest making this an optional callback of `h2o_conn_callbacks_t`. The extension needs to work fine with QUIC.

Reply:
Done in https://github.com/h2o/h2o/pull/1768/commits/0e2fab8c461704e17263c279d5bd19d163a32ecc